lib/delayed_job/threaded_worker.rb

```diff
@@ -0,0 +1,104 @@
+require 'socket'
+
+class ThreadedWorker < Delayed::Worker
+  def initialize(thread_count, options={})
+    super(options)
+    @thread_count = thread_count
+    @threads = []
+    @stop_signal = false
+    @name_prefix = options[:worker_name] || 'worker'
+  end
+
+  def start
+    trap_signals
+
+    say "Starting multi-threaded job worker with #{@thread_count} threads"
+
+    @thread_count.times do |i|
+      thread_name = generate_thread_name(i + 1)
+      @threads << Thread.new do
+        thread_name(thread_name)
+        threaded_work_off
+      end
+    end
+
+    @threads.each(&:join) # Wait for threads to finish
+  end
+
+  def name
+    Thread.current[:name] || @name
+  end
+
+  def name=(name)
+    # Set the instance variable for compatibility && ensure thread-local storage is also updated
+    @name = name
+    Thread.current[:name] = name
+  end
+
+  private
+
+  def generate_thread_name(thread_index)
+    if @name
+      "#{@name}-thread:#{thread_index}"
+    else
+      "#{@name_prefix}-host:#{Socket.gethostname} pid:#{Process.pid} thread:#{thread_index}"
+    end
+  end
+
+  def thread_name(name)
```

Comment:
Isn't this the same as `def name=(name)` above?

Reply:
The first one is the getter for `name` and the 2nd one is the setter